kong/tools/stream_api.lua

```diff
@@ -0,0 +1,100 @@
+
+require "lua_pack"
+
+
+local kong       = kong
+local st_pack    = string.pack      -- luacheck: ignore string
+local st_unpack  = string.unpack    -- luacheck: ignore string
+local st_format  = string.format
+
+local PREFIX = ngx.config.prefix()
+
+local stream_api = {}
+
+local _handlers  = {}
+
+
+function stream_api.load_handlers()
+  local utils = require "kong.tools.utils"
+
+  for plugin_name in pairs(kong.configuration.loaded_plugins) do
+    local loaded, custom_endpoints = utils.load_module_if_exists("kong.plugins." .. plugin_name .. ".api")
+    if loaded and custom_endpoints._stream then
+      kong.log.debug("Register stream api for plugin: ", plugin_name)
+      _handlers[plugin_name] = custom_endpoints._stream
+      custom_endpoints._stream = nil
+    end
+  end
+end
+
+function stream_api.request(key, data, socket_path)
+  if type(key) ~= "string" or type(data) ~= "string" then
+    kong.log.error("Key and data must be strings")
```

Comment:
you're right.  hard error and full stack dump are better for debugging.  i prefer not to include the `level` arg to `error()`, since it easily leads to suppressing the most important parts after an innocent refactoring.

error signalling is by design a slow path, and both `type()` and `assert()` are compiled away, so copying to locals doesn't buy anything.